dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/association/ProgramOrganisationUnitAssociationsQueryBuilder.java

```diff
@@ -0,0 +1,199 @@
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.association;
+
+import static java.util.stream.Collectors.joining;
+import static org.hisp.dhis.hibernate.jsonb.type.JsonbFunctions.CHECK_USER_ACCESS;
+import static org.hisp.dhis.hibernate.jsonb.type.JsonbFunctions.CHECK_USER_GROUPS_ACCESS;
+import static org.hisp.dhis.hibernate.jsonb.type.JsonbFunctions.EXTRACT_PATH_TEXT;
+import static org.hisp.dhis.hibernate.jsonb.type.JsonbFunctions.HAS_USER_GROUP_IDS;
+import static org.hisp.dhis.hibernate.jsonb.type.JsonbFunctions.HAS_USER_ID;
+import static org.hisp.dhis.security.acl.AclService.LIKE_READ_METADATA;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.commons.collection.CollectionUtils;
+import org.hisp.dhis.user.CurrentUserGroupInfo;
+import org.hisp.dhis.user.CurrentUserService;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ProgramOrganisationUnitAssociationsQueryBuilder
+{
+
+    private final CurrentUserService currentUserService;
+
+    private static final String SHARING_OUTER_QUERY_BEGIN = "select " +
+        "    prg.uid, " +
+        "    prg.agg_ou_uid " +
+        "from (";
+
+    private static final String SHARING_OUTER_QUERY_END = ") as prg";
+
+    private static final String INNER_SQL_QUERY = "select " +
+        "    pr.uid, " +
+        "    pr.sharing, " +
+        "    array_agg(ou.uid) agg_ou_uid " +
```

Comment:
Make sure we don't get duplicates in this list.

IE. if a user has access to a root OU, and multiple children of that root, the root path will also include the child ous, and the child will be duplicated.

The most performant might be to check that before the query, or you can maybe include a "distinct" somewhere?